File: solutions/leetcode/15.三数之和.cpp
/*
 * @lc app=leetcode.cn id=15 lang=cpp
 *
 * [15] 三数之和
 */

// @lc code=start
class Solution {
    public:
        vector<vector<int>> threeSum(vector<int>& nums) {
            sort(nums.begin(), nums.end());
            int n = nums.size();
    
            vector<vector<int>> ans;
            for (int i = 0; i < n; i++) {
                if (i > 0 && nums[i] == nums[i - 1]) continue; // 跳过重复的 nums[i]
                int curr = nums[i];
                if (curr > 0) break;
                int l = i + 1, r = n - 1;
                while (l < n && r >= 0 && l < r) {
                    int s = nums[i] + nums[l] + nums[r];
                    if (s == 0) {
                        ans.push_back({nums[i], nums[l], nums[r]});
                        while (l < r && nums[l] == nums[l + 1]) l++;
                        while (l < r && nums[r] == nums[r - 1]) r--;
                        l++;
                        r--;
                    }
                    else if (s > 0) r--;
                    else if (s < 0) l++;
                }
            }
            return ans;
        }
    };
// @lc code=end

File: solutions/leetcode/15.三数之和.py
#
# @lc app=leetcode.cn id=15 lang=python3
#
# [15] 三数之和
#

# @lc code=start
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        nums.sort()

        ans = []
        for i, curr in enumerate(nums):
            if i > 0 and curr == nums[i - 1]:
                continue # 跳过重复的curr
            l, r = i + 1, n - 1
            while l < r:
                s = curr + nums[l] + nums[r]
                if s == 0:
                    ans.append([curr, nums[l], nums[r]])
                    while l < r and nums[l] == nums[l + 1]:
                        l += 1
                    while l < r and nums[r] == nums[r - 1]:
                        r -= 1
                    l += 1
                    r -= 1
                elif s > 0:
                    r -= 1
                elif s < 0:
                    l += 1
        return ans
        
# @lc code=end

